reconcile/typed_queries/clusters_with_peering.py

```diff
@@ -0,0 +1,12 @@
+from reconcile.gql_definitions.common.clusters_with_peering import (
+    ClusterV1,
+    query,
+)
+from reconcile.utils import gql
+
+
+def get_clusters_with_peering() -> list[ClusterV1]:
```

Comment:
Lets add `query_func` as Callable input arg -> then we can avoid `mocker.patch`.

Reply:
E.g., https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/typed_queries/app_interface_repo_url.py#L9

Reply:
`query_func` is also already available for these tests: https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/test/test_typed_queries/conftest.py

Reply:
I didn't add `query_func` to that function is because it will lead to the branch tested in test is not the one actually used in real call. For example these [2 lines](https://github.com/app-sre/qontract-reconcile/blob/80ea1a4d6e53a3cc4fec57cb8f6c14c19538a92f/reconcile/typed_queries/app_interface_repo_url.py#L11-L12) are not test covered.
So for those functions with `query_func`, there should be a test case using `macker.patch` to test `query_func` is `None` branch, feels not worth it.

Reply:
Shouldnt these 2 lines be tested somewhere else? I dont think that `gql.get_api()` should be responsibility of the typed_query module.

Reply:
But I agree - a test for this query is not really worth it - maybe we should add more tests for the gql module instead.

Reply:
All lines inside `get_clusters_with_peering` need to be covered by `get_clusters_with_peering` unit tests, the logic of `gql.get_api().query` can be extracted to a function in `gql` like `gql.get_query_func`, then in `get_cluster_with_peering` test just need to verify `gql.get_query_func` called.

Reply:
Actually in this function, there are 2 dependencies `reconcile.gql_definitions.common.clusters_with_peering.query` and `reconcile.utils.gql`, to do it in DI way, both of them need to be injected, just not very common in python.

Reply:
You are right. Nice idea! I like the idea of `gql.get_query_func()` a lot - very clean! :heart: :+1:

Reply:
@fishi0x01 I tried to inject `query_func` to make it using DI, but mypy is not happy about it, and I feel tests are harder to understand. For example:

```python
# gql.py

P = ParamSpec("P")
R = TypeVar("R")


def inject_query_func(func: Callable[Concatenate[Callable, P], R]) -> Callable[P, R]:
    @wraps(func)
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
        query_func = get_query_func()
        return func(query_func, *args, **kwargs)

    return wrapper
```

```python
# clusters_with_peering.py

@gql.inject_query_func
def get_clusters_with_peering(query_func: Callable) -> list[ClusterV1]:
    data = query(query_func)
    clusters = data.clusters or []
    return [c for c in clusters if c.peering is not None]
```

```python
# test_clusters_with_peering.py

def test_get_clusters_with_peering(
    query_func: Callable[..., Callable],
    clusters_with_peering: ClustersWithPeeringQueryData,
) -> None:
    qf = query_func(clusters_with_peering.dict(by_alias=True))

    clusters = get_clusters_with_peering.__wrapped__(qf)

    assert clusters == clusters_with_peering.clusters
```

It works but `mypy` doesn't know `__wrapped__` (`error: "Callable[[], List[ClusterV1]]" has no attribute "__wrapped__"  [attr-defined]`), and we still need to write another test case to verify `get_clusters_with_peering` has proper `query_func` injection by `@gql.inject_query_func`.

Without a DI framework, `mocker.patch` is the only way to ensure full test coverage, I don't like to have one condition branch just used in tests, I had cases where all tests pass but still fail in real run, even that function looks like covered. What do you think about it?

Reply:
Yeah mypy and pytest are not really working together well. You will see a lot of `ignore [attr-defined]` comments in tests :sweat_smile: 

If you mock the whole gql api as you suggested above, then the following wouldnt work?

**conftest.py**
```python
@pytest.fixture
def gql() -> Callable[..., GqlApi]:
    def builder(data: Mapping) -> GqlApi:
        def query_func():
            return data
        mock = create_autospec(spec=GqlApi)
        mock.get_query_func.side_effect = [query_func]
        return mock
    return builder
```

**test.py**
```python
def test(gql, gql_class_factory):
    mock = gql(gql_class_factory(CLS, {...}).dict(by_alias=True))
    resp = my_typed_query(gql=mock)
    assert resp == ...
```

I.e., all complexity is in the `conftest.py` and the tests only use the builder function to get a `GqlApi` mock? Not 100% sure, but my understanding was that this mock can then be injected, so you still test all the lines in typed_queries module.

Reply:
Refactored with `gql_api` as dependency https://github.com/app-sre/qontract-reconcile/pull/3467/commits/a25cba1c0872677770d0e697922ee0433a8dcfcc